docs/proposals/apiserver-ha.md

```diff
@@ -0,0 +1,152 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN MUNGE: GENERATED_TOC -->
+
+- [API server HA proposal](#api-server-ha-proposal)
+    - [Introduction](#introduction)
+    - [Motivation](#motivation)
+    - [Design](#design)
+    - [Initional Implementation step](#initional-implementation-step)
+      - [1. Introduce a knob to enable quorum read](#1-introduce-a-knob-to-enable-quorum-read)
+      - [2. Default kubernetes service/endpoints](#2-default-kubernetes-serviceendpoints)
+      - [3. API server component Health check](#3-api-server-component-health-check)
+      - [4. Client behavior](#4-client-behavior)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+# API server HA proposal
+
+**Status**: Design & Implementation in progress.
+
+> Contact @mqliang or @dalanlan for questions & suggestions.
+
+### Introduction
+
+This document serves as a proposal for high availability of the API server
+component in Kubernetes. This proposal is intended to provide a simple high
+availability solution for API server component.
+
+### Motivation
+
+Currently, high availability solution for scheduler and controller components
+in Kubernetes has been proposed, and are under heavy development. The HA
+solution for scheduler and controller components are Warm Standby, which
+mean there is only one active component acting as the master and additional
+components running but not providing service or responding to requests.
+
+API server should also support HA so that k8s will be a truly reliable, highly
+available distributed system. As a matter fact, API server is already highly
+available in some sense since it's stateless, and we could build up a k8s cluster
+with replicated API servers easily(see
+[high-availability](../../docs/admin/high-availability.md) ).
+However, there are still some problems
+need to address. This proposal aims at address those problems.
+
+### Design
+
+```
+  |--------|            |--------|            |--------|              
```

Comment:
@timothysc I would have to think about it. My initial thought is that I'm not sure what you'd really gain by doing that.